docs/content/guides/dagster/transitioning-data-pipelines-from-development-to-production.mdx

```diff
@@ -0,0 +1,456 @@
+                    "user": "system@company.com",
+                    "password": {"env": "SYSTEM_SNOWFLAKE_PASSWORD"},
+                    "database": "PRODUCTION",
+                    "schema": "HACKER_NEWS",
+                }
+            ),
+        },
+    }
+    deployment_name = os.getenv("DAGSTER_DEPLOYMENT", "local")
+
+    return [
+        *with_resources(
+            [items, comments, stories], resource_defs=resource_defs[deployment_name]
+        )
+    ]
+```
+
+### Staging
+
+Depending on your organization’s Dagster setup, there are a couple of options for a staging environment.
+
+- **For Dagster Cloud users**, we recommend using [Branch Deployments](https://docs.dagster.cloud/guides/branch-deployments) as your staging step. A Branch Deployment is a new Dagster deployment that is automatically generated for each git branch. Check out our \[comprehensive guide to Branch Deployments]\(link to claire’s doc) to learn how to use Branch Deployments to verify data pipelines before deploying them to production.
+
+- **For a self-hosted staging deployment**, we’ve already done most of the necessary work to run our assets in staging! All we need to do is add another entry to the `resource_defs` dictionary in our repository and set `DAGSTER_DEPLOYMENT=staging` in our staging deployment.
+
+```python file=/guides/dagster/development_to_production/repository/repository_v3.py startafter=start_staging endbefore=end_staging
+resource_defs = {
+    "local": {...},
+    "production": {...},
+    "staging": {
+        "snowflake_io_manager": snowflake_io_manager.configured(
+            {
+                "account": "abc1234.us-east-1",
+                "user": "system@company.com",
+                "password": {"env": "SYSTEM_SNOWFLAKE_PASSWORD"},
+                "database": "STAGING",
+                "schema": "HACKER_NEWS",
+            }
+        ),
+    },
+}
+```
+
+---
+
+## Optional: Testing
+
+You may have noticed a missing step in the development workflow presented in this guide — unit tests! While the main purpose of the guide is to help you transition your code from local development to a production deployment, unit testing is still an important part of the development cycle. In this section, we'll explore a pattern you may find useful when writing your own unit tests.
+
+When we write unit tests for the `items` asset, we could make more precise assertions if we knew exactly what data we'd receive from Hacker News. If we refactor our interactions with the Hacker News API as a resource, we can leverage Dagster's resource system to provide a stub resource in our unit tests.
+
+Before we get into implementation, let's go over some best practices:
+
+### When to use resources
+
+In many cases, it is often most convenient to interact with an external service directly in assets or ops, rather than write a resource. We recommend refactoring code to use resources in the following cases:
```

Comment:
Just to make the relationship between resources and external services super explicit, I think it would be helpful to change the first sentence to be something like: it is often most convenient to interact with an external service directly in assets or ops, rather than refactoring the code that interacts with the service to exist within a resource.